File: 2017/day15/part1.scala
class generator(var current: Long, val factor: Long) {
  def next: Long = {
    current = (current * factor) % 2147483647
    current
  }
}
val a = new generator(65, 16807)
val b = new generator(8921, 48271)

var total = 0
for (i <- 1 to 40000000) {
//  if (BigInt(a.next).toByteArray.takeRight(2) sameElements BigInt(b.next).toByteArray.takeRight(2)) {
  if (a.next % 65536 == b.next % 65536) {
    total += 1
  }
}
println(total)

File: 2017/day15/part1.py
def gen(n, factor):
    while True:
        n = (n* factor) % 2147483647
        yield n

def b16_match(a, b):
    return ( a % 256 == b % 256 and          # First byte
             a / 256 % 256 == b / 256 % 256) # Second byte

a = gen(int(raw_input("Generator A: [65] ") or 65), 16807)
b = gen(int(raw_input("Generator B: [8921] ") or 8921), 48271)

total = 0
for i in xrange(0,40000000):
    if b16_match(a.next(), b.next()):
        total += 1
print total
